Synthesising BitPack instances for type with phantom parameter fails

I've been having trouble synthesising the following code (in Clash 1.2.0) which uses types which include phantom type parameters being included in a record.
```
{-# LANGUAGE StandaloneDeriving    #-}
{-# LANGUAGE UndecidableInstances  #-}
module Project where

import           Clash.Prelude
import           Control.DeepSeq   (NFData)
import           GHC.Generics      (Generic)

data RepKind = AppRep | WireRep
  deriving (Generic, Show, Eq, NFData, NFDataX)

newtype U16 (t :: RepKind) = U16 (Unsigned 16)

deriving instance Generic (U16 a)
deriving instance Eq (U16 a)
deriving instance Show (U16 a)
deriving instance NFData (U16 a)
deriving instance NFDataX (U16 a)
deriving instance BitPack (U16 'AppRep)

class WireApp w a where
    toWire :: a -> w
    toApp :: w -> a

instance WireApp (U16 'WireRep) (U16 'AppRep) where
    toWire v = case v of U16 x -> U16 x
    toApp v = case v of U16 x -> U16 x

instance (BitPack (t 'AppRep), WireApp (t 'WireRep)  (t 'AppRep)) => BitPack (t 'WireRep) where
  type BitSize (t 'WireRep) = BitSize (t 'AppRep)
  pack x = bv
    where
        bv :: BitVector (BitSize (t 'WireRep))
        bv = pack app
        app :: t 'AppRep
        app = toApp $ x
  unpack x = toWire app
    where
        app :: t 'AppRep
        app = unpack x

data Record
  = Record
  { f1 :: U16 'WireRep
  , f2 :: U16 'WireRep
  } deriving (Generic, NFData, Show, Eq, BitPack, NFDataX)

topEntity :: Signal System Bool -> Signal System Bool
topEntity _ = pure (unpack 0 == Record (U16 0) (U16 0))
```

When I attempt to synthesise this code to VHDL I am getting the below error.
When I remove the `--vhdl` flag, compilation completes fine.
Is there any known reason for this?

```
$ stack exec clash -- -XBinaryLiterals -XDataKinds -XDeriveAnyClass -XDeriveGeneric -XDeriveLift -XDuplicateRecordFields -XFlexibleContexts -XFlexibl
eInstances -XFunctionalDependencies -XLambdaCase -XMagicHash -XMultiParamTypeClasses -XMultiWayIf -XScopedTypeVariables -XTemplateHaskell -XTypeApp
lications -XTypeFamilies -XTypeOperators -XNoImplicitPrelude -XTypeSynonymInstances -XTupleSections -fconstraint-solver-iterations=100 -fplugin=GHC
.TypeLits.KnownNat.Solver -fplugin=GHC.TypeLits.Extra.Solver -fplugin=GHC.TypeLits.Normalise --vhdl -fclash-error-extra -fclash-compile-ultra ./Pro
ject.hs
WARNING: Clash is linked statically, which can lead to long startup times.
See https://gitlab.haskell.org/ghc/ghc/issues/15524

GHC: Parsing and optimising modules took: 1.802s
GHC: Loading external modules from interface files took: 0.454s
GHC: Parsing annotations took: 0.001s
Clash: Parsing and compiling primitives took 0.066s
GHC+Clash: Loading modules cumulatively took 2.539s
Clash: Compiling Project.topEntity
Clash: Normalisation took 0.010s

Project.hs:49:1: error:

    Clash.Netlist(277): Clash.Core.Util(844): Cannot reduce to an integer:
    Clash.Class.BitPack.BitSize[8214565720323789400]
      (Project.U16[8214565720323850163]
         Project.WireRep[8214565720323850167])

    The source location of the error is not exact, only indicative, as it is acquired
    after optimizations. The actual location of the error can be in a function that is
    inlined. To prevent inlining of those functions, annotate them with a NOINLINE pragma.
   |
49 | topEntity _ = pure (unpack 0 == Record (U16 0) (U16 0))
   | ^^^^^^^^^
make: *** [clash-example] Error 1
```

--- a/changelog/2020-03-24T13:08:40+01:00_fix1242
+++ b/changelog/2020-03-24T13:08:40+01:00_fix1242
@@ -1,0 +1,1 @@
+FIXED: [#1242](https://github.com/clash-lang/clash-compiler/issues/1242) Synthesizing BitPack instances for type with phantom parameter fails

--- a/clash-lib/src/Clash/Core/Type.hs
+++ b/clash-lib/src/Clash/Core/Type.hs
@@ -453,10 +453,8 @@ funSubst tcm (Just s) = uncurry go
       Just ty' | ty' `aeqType` ty -> Just s
       _ -> Nothing
     go ty1 (reduceTypeFamily tcm -> Just ty2) = go ty1 ty2 -- See [Note: lazy type families]
-    go ty1@(LitTy _) ty2 = if ty1 `aeqType` ty2 then Just s else Nothing
-    go (tyView -> TyConApp tc argTys) (tyView -> TyConApp tc' argTys')
-      | tc == tc'
-      = foldl' (funSubst tcm) (Just s) (zip argTys argTys')
+    -- [Note] funSubst FunTy
+    --
     -- Whenever type classes have associated types whose instances 'map' to
     -- functions, we try to find substitutions in the LHS and RHS of these
     -- (type-level) functions. Because we use @funSubst@ recursively, we
@@ -473,9 +471,17 @@ funSubst tcm (Just s) = uncurry go
     -- matching against `Char -> Char` we'd find a duplicate `a -> Char`. We
     -- can't think of any (type-checking) cases where these mappings would map
     -- to different types, so this is OK for our purposes.
-    go (tyView -> FunTy a b) (tyView -> FunTy a' b') =
-      (++) <$> funSubst tcm (Just s) (a, a')
-           <*> funSubst tcm (Just s) (b, b')
+    go (AppTy a1 r1) (AppTy a2 r2) = do
+      s1 <- funSubst tcm (Just s) (a1, a2)
+      funSubst tcm (Just s1) (r1, r2)
+
+    go ty1@(ConstTy _) ty2 =
+      -- Looks through AnnType
+      if ty1 `aeqType` ty2 then Just s else Nothing
+
+    go ty1@(LitTy _) ty2 =
+      -- Looks through AnnType
+      if ty1 `aeqType` ty2 then Just s else Nothing
 
     go _ _ = Nothing
 

--- a/tests/shouldwork/Basic/T1242.hs
+++ b/tests/shouldwork/Basic/T1242.hs
@@ -1,0 +1,50 @@
+{-# LANGUAGE StandaloneDeriving    #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+module T1242 where
+
+import           Clash.Prelude
+import           Control.DeepSeq   (NFData)
+import           GHC.Generics      (Generic)
+
+data RepKind = AppRep | WireRep
+  deriving (Generic, Show, Eq, NFData, NFDataX)
+
+newtype U16 (t :: RepKind) = U16 (Unsigned 16)
+
+deriving instance Generic (U16 a)
+deriving instance Eq (U16 a)
+deriving instance Show (U16 a)
+deriving instance NFData (U16 a)
+deriving instance NFDataX (U16 a)
+deriving instance BitPack (U16 'AppRep)
+
+class WireApp w a where
+    toWire :: a -> w
+    toApp :: w -> a
+
+instance WireApp (U16 'WireRep) (U16 'AppRep) where
+    toWire v = case v of U16 x -> U16 x
+    toApp v = case v of U16 x -> U16 x
+
+instance (BitPack (t 'AppRep), WireApp (t 'WireRep)  (t 'AppRep)) => BitPack (t 'WireRep) where
+  type BitSize (t 'WireRep) = BitSize (t 'AppRep)
+  pack x = bv
+    where
+        bv :: BitVector (BitSize (t 'WireRep))
+        bv = pack app
+        app :: t 'AppRep
+        app = toApp $ x
+  unpack x = toWire app
+    where
+        app :: t 'AppRep
+        app = unpack x
+
+data Record
+  = Record
+  { f1 :: U16 'WireRep
+  , f2 :: U16 'WireRep
+  } deriving (Generic, NFData, Show, Eq, BitPack, NFDataX)
+
+topEntity :: Signal System Bool -> Signal System Bool
+topEntity _ = pure (unpack 0 == Record (U16 0) (U16 0))

--- a/testsuite/Main.hs
+++ b/testsuite/Main.hs
@@ -266,6 +266,7 @@ runClashTest = defaultMain $ clashTestRoot
           , topEntities=TopEntities ["top1"]
           }
         , runTest "T1012" def{hdlSim=False}
+        , runTest "T1242" def{hdlSim=False}
         , runTest "TagToEnum" def{hdlSim=False}
         , runTest "TestIndex" def{hdlSim=False}
         , runTest "Time" def

